### PR TITLE
return missing headers when raising HeaderRowNotFoundError

### LINKED ISSUE #412: Suggestion: have HeaderRowNotFoundError report which header is missing
Right now if you call #parse with a hash on a spreadsheet with header that isn't found in the sheet it raises `Roo::HeaderRowNotFoundError`

for example if you have a simple spreadsheet like such
``` 
header1, header2
one,two
```

and you call  
``` rb
Roo::Spreadsheet.open(FILE_PATH).parse(header1: 'header1', header3: 'header3') 
````

It will raise `Roo::HeaderRowNotFoundError` but there doesn't appear to be a way to easily know which header (in this case "header3") is missing from the spreadsheet. I find the HeaderRowNotFoundError helpful when debugging big spreadsheets with many columns, but it would be even more useful to know which header was missing from the sheet!

### PR DESCRIPTION
**Pull request by bummfelix**
### Summary

return missing headers when raising HeaderRowNotFoundError
fixes #412

### CODE REVIEW

File: spec/lib/roo/base_spec.rb [hunk 1]
```diff
@@ -154,9 +154,14 @@ def sheets
...
-      it 'raises an error' do
+      it 'raises an error ' do
         expect { spreadsheet.row_with([/Missing Header/]) }.to \
-          raise_error(Roo::HeaderRowNotFoundError)
+          raise_error(Roo::HeaderRowNotFoundError, /Missing Header/)
+      end
+
+      it 'returns missing headers' do
+        expect { spreadsheet.row_with([/Header/, /Missing Header 1/, /Missing Header 2/]) }.to \
+          raise_error(Roo::HeaderRowNotFoundError, /Missing Header 1|Missing Header 2/)
```
Comment (chopraanmol1) [inline spec/lib/roo/base_spec.rb]: `/Missing Header 1|Missing Header 2/` only ensures that one of missing header is part of exception. 

If I change the code to raise just `missing_headers.first` this test case will still pass. I think this regex should ensure that all missing header/query is part of exception.

File: spec/lib/roo/base_spec.rb [hunk 2]
```diff
@@ -154,9 +154,14 @@ def sheets
     end
 
     context 'without a matching header row' do
-      it 'raises an error' do
+      it 'raises an error ' do
```
Comment (chopraanmol1) [inline spec/lib/roo/base_spec.rb]: `'raises an error ' ` -> `'raises an error' `

File: lib/roo/base.rb [hunk 3]
```diff
@@ -308,18 +308,23 @@ def parse(options = {})
...
-        raise Roo::HeaderRowNotFoundError
+      else
+        closest_mismatched_headers = headers if headers.length > closest_mismatched_headers.length
+        if line_no > 100
+          break
+        end
       end
     end
-    raise Roo::HeaderRowNotFoundError
+    missing_headers = query.each_with_object([]) \
```
[the remark this answers is not in the document]
Comment (bummfelix) [reply lib/roo/base.rb]: agree, select is prettier

File: spec/lib/roo/base_spec.rb (hunk 1 above)
Comment (bummfelix) [reply spec/lib/roo/base_spec.rb]: agree, changed that!

File: spec/lib/roo/base_spec.rb (hunk 2 above)
Comment (bummfelix) [reply spec/lib/roo/base_spec.rb]: changed

File: spec/lib/roo/base_spec.rb [hunk 4]
```diff
@@ -158,6 +158,11 @@ def sheets
         expect { spreadsheet.row_with([/Missing Header/]) }.to \
           raise_error(Roo::HeaderRowNotFoundError)
       end
+
+      it 'returns missing headers' do
+        expect { spreadsheet.row_with([/Header/, /Missing Header 1/, /Missing Header 2/]) }.to \
+          raise_error(Roo::HeaderRowNotFoundError, /(?=.*Missing Header 1)(?=.*Missing Header 2)/)
```
Comment (chopraanmol1) [inline spec/lib/roo/base_spec.rb]: This test will pass even if I raise `query` instead of `missing_headers`

Comment (bummfelix) [reply spec/lib/roo/base_spec.rb]: Totally true, decided to check for returned string, improves readability imho

**Pull request merged**